### PR TITLE
[generate_vmap_rule] Add generate_vmap_rule to autograd.Function

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91211
* #91206
* __->__ #90966
* #90965
* #90964
* #90963

Design document:
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit

This PR adds a `generate_vmap_rule` option (default False) to autograd.Function.
By setting it to True, a user promises to us that their autograd.Function's
{forward, backward, jvp}, if defined, only uses PyTorch operations, in addition to the other
limitations of autograd.Function+functorch (such as the user not
capturing any Tensors being transformed over from outside of the
autograd.Function).

Concretely, the approach is:
- we update `custom_function_call` to accept an additional
`generate_vmap_rule` argument.
- The vmap rule for `custom_function_call` and `generate_vmap_rule=True`
is: we construct a vmapped version of the autograd.Function and dispatch
on it.
- The vmapped version of the autograd.Function can be thought of like
the following: if we have an autograd.Function Foo, then
VmappedFoo.apply(in_dims, ...) has the same semantics as
vmap(Foo.apply, in_dims...)
- VmappedFoo's forward, setup_context, and backward staticmethod are
vmapped versions of Foo's staticmethods.
- See the design doc for more motivation and explanation

Test Plan:
- This PR introduces additional autograd.Function with the suffix "GenVmap" to
autograd_function_db.
- There are also some minor UX tests

Future:
- jvp support
- likely more testing to come, but please let me know if you have
cases that you want me to test here.

### CODE REVIEW

File: torch/testing/_internal/autograd_function_db.py [hunk 1]
```diff
@@ -267,11 +366,83 @@ def vmap(info, in_dims, x, idx):
...
+        )
+
+    @staticmethod
+    def jvp(ctx, gx, gy):
+        # Intentionally returning torch.zeros instead of zeros_like or new_zeros.
+        # Also intentionally not None.
+        return (
+            torch.zeros(gx.shape, dtype=gx.dtype, device=gx.device),
+            torch.zeros(gy.shape, dtype=gy.dtype, device=gy.device),
+        )
```
Comment (soulitzer) [inline torch/testing/_internal/autograd_function_db.py:445]: A little unrelated, but maybe we also want to a custom Function where the gradient is intentionally wrong for the case where vmap is provided by the user, just to make sure the backward is preserved when we do grad(vmap for that as well. Currently if the user provided vmap does not call back into the original custom Function it seems possible that the backward will not be preserved.

Comment (zou3519) [reply torch/testing/_internal/autograd_function_db.py:445]: > Currently if the user provided vmap does not call back into the original custom Function it seems possible that the user-provided backward will not be preserved.

In this case, the user-provided backward is not preserved if the user provides `vmap` and does not call back into the original custom Function. This is intentional:
- this is consistent with how aten operators work when we do grad(vmap(f))(x): let's say f is at::dot and vmap transforms at::dot into at::mm. the grad records an at::mm node, and MMBackward gets run instead of a vmapped DotBackward.
- it's on the user to provide a correct `vmap`

File: torch/_functorch/autograd_function.py [hunk 2]
```diff
@@ -240,11 +252,167 @@ def custom_function_call_vmap(interpreter, autograd_function, *operands):
...
+                                      True,  # generate_vmap_rule
+                                      *unwrapped_operands)
+
+    out_dims = get_out_dims()
+    return wrap_batched(output, out_dims, interpreter.level())
+
+
 @custom_function_call.py_impl(TransformType.Functionalize)
-def custom_function_call_functionalize(interpreter, autograd_function, *operands):
+def custom_function_call_functionalize(interpreter, autograd_function, generate_vmap_rule, *operands):
```
Comment (soulitzer) [inline torch/_functorch/autograd_function.py:270]: Instead of having to pass this, could we just check its value on the autograd_function?

File: test/functorch/test_ops.py [hunk 3]
```diff
@@ -1517,6 +1519,9 @@ def reference(primals, cotangents, primals_tangents, cotangents_tangents):
         xfail("_native_batch_norm_legit"),
         xfail('native_dropout_backward'),
         xfail('nn.functional.prelu'),
+
+        xfail('CubeGenVmapAutogradFunction'),  # NYI
+        xfail('SortGenVmapAutogradFunction'),  # https://github.com/pytorch/pytorch/issues/90067
```
Comment (soulitzer) [inline test/functorch/test_ops.py:1524]: This has been fixed now! https://github.com/pytorch/pytorch/issues/90067

**Review by soulitzer — COMMENTED** [review]
Looks mostly good, just a few comments

File: aten/src/ATen/functorch/Interpreter.cpp [hunk 4]
```diff
@@ -88,10 +89,10 @@ void sanityCheckStack(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
   auto num_args = op.schema().arguments().size();
   foreachTensorInplace(*stack, stack->size() - num_args, stack->size(),
       [](const Tensor& tensor) {
-
-        auto* wrapper = maybeGetTensorWrapper(tensor);
+        auto result = unwrapIfDead(tensor);
+        auto* wrapper = maybeGetTensorWrapper(result);
         TORCH_INTERNAL_ASSERT(wrapper == nullptr);
-        auto* batched = maybeGetBatchedImpl(tensor);
+        auto* batched = maybeGetBatchedImpl(result);
```
[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/functorch/Interpreter.cpp:95]: This can happen if someone invokes a functorch transform on a dead GradTensorWrapper and ends up wrapping the dead GradTensorWrapper in an alive wrapper. Some situations where it can happen are:
- if we do vjp(vmap over an autograd.Function with generate_vmap_rule=True
- if we do vjp( over an autograd.Function with a vmap staticmethod AND the backward pass calls a functorch transform. vjp saves GradTensorWrapper for backward, so in the backward pass, all of those GradTensorWrapper are dead.

This doesn't happen if we do vjp(vmap over a regular PyTorch operator, because the implementation of the operator doesn't end up calling a functorch transform.

File: torch/_functorch/autograd_function.py (hunk 2 above)
Comment (zou3519) [reply torch/_functorch/autograd_function.py:270]: Yeah, that's a better idea, thanks for the suggestion

File: test/functorch/test_ops.py (hunk 3 above)
Comment (zou3519) [reply test/functorch/test_ops.py:1524]: Thank you, that was quick

File: torch/_functorch/autograd_function.py [hunk 5]
```diff
@@ -274,6 +442,50 @@ def __setattr__(self, name, value):
...
+        self._pt_saved_tensors_bdims = bdims
+
+    def __getattr__(self, attr):
+        return getattr(self._pt_inner_ctx, attr)
+
+    def __setattr__(self, attr, value):
+        if attr in ctx_custom_save_properties:
+            self.__dict__[attr] = value
+            return
+        return setattr(self._pt_inner_ctx, attr, value)
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_functorch/autograd_function.py:473]: These ctx wrappers do different things so I would prefer to keep them separate. This one has custom save_for_backward behavior, the other one is used for overriding ctx.saved_tensors.

The similarity is in how they're both context wrappers that forward attr accesses to the underlying ctx object. I can make a common base class that both of these inherit from to avoid repetition, if that sounds reasonable

[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/_functorch/autograd_function.py:473]: Sounds good!

File: aten/src/ATen/functorch/Interpreter.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply aten/src/ATen/functorch/Interpreter.cpp:95]: Makes sense, thanks for clarifying!

File: test/functorch/test_ops.py (hunk 3 above)
Comment (zou3519) [reply test/functorch/test_ops.py]: Updated to NYI (this PR doesn't add jvp support for generate_vmap_rule=True)

File: torch/_functorch/autograd_function.py (hunk 2 above)
Comment (zou3519) [reply torch/_functorch/autograd_function.py:268]: Updated

File: torch/_functorch/autograd_function.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_functorch/autograd_function.py]: Updated

File: torch/_functorch/autograd_function.py [hunk 6]
```diff
@@ -240,11 +252,165 @@ def custom_function_call_vmap(interpreter, autograd_function, *operands):
...
 
+def vmapify_autograd_function(autograd_function, in_dims, batch_size, randomness):
+    # The following values are saved from the forward() and setup_context()
+    # and used in backward().
+    # Why do we save the values out here instead of on the ctx object?
+    # - out_dims: There's no way to retrieve this from forward()
+    # - input_shapes, saved_tensors_bdims: I'm a bit scared of nesting
+    #   vmap(vmap( but not completely sure if it is a problem. If we
+    #   assigned those fields to the ctx object, the worry is that they
+    #   get overwritten.
```
Comment (soulitzer) [inline torch/_functorch/autograd_function.py:280]: For input_shapes and saved_tensor_bdims, isn't it less likely to get overridden when you do `grad(grad(vmap` if you save them on ctx instead of here because you make a new ctx for every layer of grad you have.

Currently I would think that it is getting overridden because the two grads refer to the same vmapified autograd_function,  but it just doesn't matter because both grad layers should be saving the same input_shapes + saved-tensor_bdims (because there's no vmap in between). And even if you have something like `grad(vmap(grad(vmap`, that doesn't matter because now the two grads should be referring to different autograd_function objects.

File: test/functorch/test_ops.py (hunk 3 above)
Comment (zou3519) [reply test/functorch/test_ops.py]: By the way, I'm not immediately to remove the xfail in https://github.com/pytorch/pytorch/blob/f02e93b584ff77519af10f66ad5cfa644da19428/test/functorch/test_ops.py#L1360 after your PR.

The reason being is that it looks like it creates an torch.int64 tangent tensor instead of passing None: https://github.com/pytorch/pytorch/blob/f02e93b584ff77519af10f66ad5cfa644da19428/torch/testing/_internal/autograd_function_db.py#L238-L239 .

ctx.set_materialize_grads does make the problem go away, but... (1) morally I feel like there should never be a non-fp tangent tensor and (2) the set_materialize_grads documentation doesn't say anything about this case

File: torch/_functorch/autograd_function.py (hunk 6 above)
Comment (zou3519) [reply torch/_functorch/autograd_function.py:280]: If you do grad(grad(vmap, then yeah, those are not being overridden because each grad layer gets a different ctx object.

The concern is for two consecutive vmap, like grad(vmap(vmap. I haven't actually tested this so it *might* just work out of the box, but another benefit of keeping these attributes out here is that I don't need to denylist more attrs for the user. Though there are other ways around the denylist (we can shove all extra attrs we want into a single object on the ctx)

File: test/functorch/test_ops.py (hunk 3 above)
Comment (soulitzer) [reply test/functorch/test_ops.py]: Yeah that looks l like a bug, taking a look.

Comment (soulitzer) [reply test/functorch/test_ops.py]: Should be fixed by https://github.com/pytorch/pytorch/pull/91183

**Review by soulitzer — APPROVED** [review]
LGTM
